### PR TITLE
fix px4 launch file for ROS2

### PR DESCRIPTION
**Pull request by vacabun**
I modified px4 launch file for ROS2, and tested it in the simulation environment.

### CODE REVIEW

**Review by vooon — COMMENTED** [review]
Overally looks good. Can you please do the same for apm?

File: mavros/launch/px4_config.yaml [hunk 1]
```diff
@@ -138,144 +152,160 @@ mission:
...
+      horizontal_fov_ratio: 1.0   # horizontal_fov = horizontal_fov_ratio * msg.field_of_view
+      vertical_fov_ratio: 1.0     # vertical_fov = vertical_fov_ratio * msg.field_of_view
+      # custom_orientation:       # Used for orientation == CUSTOM
+      #   roll: 0
+      #   pitch: 270
+      #   yaw: 0
+    laser_1_sub:
+      subscriber: true
+      id: 3
+      orientation: PITCH_270
```
[the remark this answers is not in the document]
Comment (vacabun) [reply mavros/launch/px4_config.yaml]: Ok, I see

**Comment by vacabun** [comment]
OK, I will try to do the same for apm.

**Comment by vooon** [comment]
Yes, bridge is not needed anymore, because you have router.
And I've removed event_launcher as it should be done differently, like a plugin to launch system, or something like that.

**Comment by vooon** [comment]
Fix #1564
Replace #1747

**Comment by vacabun** [comment]
OK，I have removed them.

File: mavros/launch/px4_config.yaml [hunk 2]
```diff
@@ -44,87 +47,98 @@ cmd:
...
+    frame_id: "map"             # origin frame
+    child_frame_id: "base_link" # body-fixed frame
+    rot_covariance: 99999.0   # covariance for attitude?
+    gps_uere: 1.0             # User Equivalent Range Error (UERE) of GPS sensor (m)
+    use_relative_alt: true    # use relative altitude for local coordinates
+    tf:
+      send: false               # send TF?
+      frame_id: "map"  # TF frame_id
+      global_frame_id: "earth"  # TF earth frame_id
+      child_frame_id: "base_link" # TF child_frame_id
```
Comment (vooon) [inline mavros/launch/px4_config.yaml:61]: In ROS2 parameters are flat, so i had to change names. Please have a look:

https://github.com/mavlink/mavros/blob/6e86500318385b5770653c6e47e3c0cdda2e3181/mavros/src/plugins/global_position.cpp#L90-L107

File: mavros/launch/px4_config.yaml [hunk 3]
```diff
@@ -44,87 +47,98 @@ cmd:
...
 
 # rc_io
 # None
 
-# safety_area
-safety_area:
-  p1: {x:  1.0, y:  1.0, z:  1.0}
-  p2: {x: -1.0, y: -1.0, z: -1.0}
+# # safety_area
+/mavros/safety_area:
```
Comment (vooon) [inline mavros/launch/px4_config.yaml:90]: Safety area is quite deprecated feature. I've removed it from ROS2.

File: mavros/launch/px4_config.yaml [hunk 4]
```diff
@@ -138,144 +152,161 @@ mission:
...
-    #   roll: 0
-    #   pitch: 270
-    #   yaw: 0
-  laser_1_sub:
-    subscriber: true
-    id: 3
-    orientation: PITCH_270
+/mavros/distance_sensor:
+  ros__parameters:
+    config: |
```
Comment (vooon) [inline mavros/launch/px4_config.yaml:157]: Yes, that's what i mean. +1

File: mavros/launch/px4_config.yaml [hunk 5]
```diff
@@ -138,144 +152,161 @@ mission:
...
-    rate_limit: 10.0      # TF rate
-  gps_rate: 5.0           # GPS data publishing rate
+/mavros/fake_gps:
+  ros__parameters:
+    # select data source
+    use_mocap: true         # ~mocap/pose
+    mocap_transform: true   # ~mocap/tf instead of pose
+    use_vision: false       # ~vision (pose)
+    # origin (default: Zürich)
+    geo_origin:
```
Comment (vooon) [inline mavros/launch/px4_config.yaml:201]: https://github.com/mavlink/mavros/blob/6e86500318385b5770653c6e47e3c0cdda2e3181/mavros_extras/src/plugins/fake_gps.cpp#L133-L144

So it should look like:
```
fake_gps:
  ros__parameters:
    geo_origin.lat: 47.3667
```

**Review by vooon — COMMENTED** [review]
Have you tried to use this launch?
I just have an idea, that params file not fully loaded.

Do you know, is it possible to define namespace for params?
Just if you run several copies of the UAS to control few drones, it might be convenient to have same config.

File: mavros/launch/px4_config.yaml (hunk 5 above)
Comment (vacabun) [reply mavros/launch/px4_config.yaml:201]: OK, I see.

File: mavros/launch/px4_config.yaml (hunk 2 above)
Comment (vacabun) [reply mavros/launch/px4_config.yaml:61]: OK, I see.

File: mavros/launch/px4_config.yaml (hunk 3 above)
Comment (vacabun) [reply mavros/launch/px4_config.yaml:90]: OK, I see.

**Comment by vacabun** [comment]
I have used this launch, but just tested a little. 

I think is a good idea. That will be really convenient for using multiple drones.

**Comment by vacabun** [comment]
The plugin `odometry` in apm launch has two parameters.

https://github.com/mavlink/mavros/blob/a55c747cf8386384bb2c7c2eb56bb62f0b6bb2c8/mavros/launch/apm_config.yaml#L228-L232

I can't find the parameters `frame_tf.desired_frame` and `estimator_type` in `mavros_extras/src/plugins/odom.cpp`. 
Should I modify it as px4 launch.

https://github.com/mavlink/mavros/blob/a55c747cf8386384bb2c7c2eb56bb62f0b6bb2c8/mavros/launch/px4_config.yaml#L232-L235

Finally, it looks like:
```
# odom
/mavros/odometry:
  ros__parameters:
    fcu.odom_parent_id_des: "map"    # desired parent frame rotation of the FCU's odometry
    fcu.odom_child_id_des: "base_link"    # desired child frame rotation of the FCU's odometry
```

**Comment by vooon** [comment]
Odom have changed a lot, so it's not a surprise that conf outdated.

We should use parameters what ROS2 have.
https://github.com/mavlink/mavros/blob/6e86500318385b5770653c6e47e3c0cdda2e3181/mavros_extras/src/plugins/odom.cpp#L67-L74

**Review by vooon — APPROVED** [review]
LGTM, Thanks!

**Comment by vooon** [comment]
Yes, in ROS2 i've implemented idea of splitting mavlink connections and UAS. So if previously you have to run several copies of whole mavros_node, and maybe gcs_nodes, now it utilizes ros2 loadable node classes.

```
FCU1 -> Router -> UAS1
FCU2 --^          &-> UAS2
                  &-> GCS
```

**Pull request merged**